### PR TITLE
Add ability to "init" services

### PR DESCRIPTION
**Pull request by acron0**
Services that implement "IInitialise" will now be initialised when venue starts. Also added a test to check this.

### CODE REVIEW

File: example/src/example/services/data.cljs [hunk 1]
```diff
@@ -1,8 +1,22 @@
...
+  (fn [result response cursor] result))
 
-(defmethod handler
-  :login
-  [_ args]
-  [:success {:token "foobar"}])
+(defmulti request-handler
+  (fn [event args response-ch] event))
+
+(defn service
```
Comment (sw1nn) [inline example/src/example/services/data.cljs]: I think you want a deftype here. These anonymous classes are hard to debug.

See https://github.com/cemerick/clojure-type-selection-flowchart

File: src/venue/core.cljs [hunk 2]
```diff
@@ -120,6 +120,13 @@
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
+(defn- init-services
+  []
+  (doseq [service (:services @state)]
+    (let [handler ((val service))]
+      (when (satisfies? IInitialise handler)
```
Comment (sw1nn) [inline src/venue/core.cljs]: The preference here is to remove the `satisfies?` check but extend the protocol to java.lang.Object and provide a no-op initialise fn. You can also extend the protocol to nil, which can be useful, but consider if it's better to throw the NPE in that case (it depends).

File: src/venue/core.cljs [hunk 3]
```diff
@@ -239,13 +246,16 @@
 
 (defn- add-service!
   [{:keys [id handler]}]
-  (if (satisfies? IHandleRequest (handler))
-    (swap! state assoc-in [:services id] handler)
-    (log-severe "Services must implement IHandleRequest." id))
+  (if-not handler
+    (log-severe "Service is null." id)
+    (do
+      (if (satisfies? IHandleRequest (handler))
```
Comment (sw1nn) [inline src/venue/core.cljs]: similar to above, may not be possible if you really need the logging. not really grokking this so happy to be corrected.

File: src/venue/core.cljs (hunk 2 above)
Comment (acron0) [inline src/venue/core.cljs]: I want to talk about this more. I added a comment to the issue you raised regarding view-models using `reify` which is philosophically identical to this: https://github.com/MastodonC/venue/issues/5

**Pull request merged**